### PR TITLE
4-Create App Flavors for Dev and Prod (flutter_flavor)

### PR DESCRIPTION
**Pull request by cxw320**
### **Summary:** 
This PR adds in 2 app flavors. One is called "dev" and can be re-purposed for setting-up a staging api. One is called "prod" and can be re-purposed to set up a production api. 

### **How to run and test this:**
There are 2 ways to run flavors. 

**1) In VSCode**, if you click on the Run and Debug pane, you'll see a drop-down with different build configurations. You can select "dev" and "prod" flavors via this drop-down. 
Screenshot 2023-10-20 at 3 48 48 PM

**2. In the terminal**, you can run the below commands:

**For Prod:**
`flutter run --flavor prod -t lib/main_prod.dart`

**For Dev:**
`flutter run --flavor dev -t lib/main_dev.dart`

If you run them successfully, you should see the below UI:

**For Prod: (you should see the red banner in top right say prod)**
Screenshot 2023-10-20 at 3 51 59 PM

**For Dev: (you should see the red banner in top right say dev)**
Screenshot 2023-10-20 at 3 55 47 PM

### **Steps taken to Implement this**

This was implemented following guidelines from the below articles:
https://itnext.io/flutter-new-app-setup-with-flavors-in-one-go-331471b127e3  (great resource for setting up in iOS and Android)
https://pub.dev/packages/flutter_flavor#flutter_flavor (how to set up FlavorConfig in dart)
https://medium.com/@devrient.jonas/flutter-profiles-in-vscode-product-flavors-c95231818005 (setting up flavor profiles to be launched in VS Code)

**Step 1) Add build flavors in Android**
Add flavors to app/build.gradle. The resValue can be used to define any type of resource, but in the case below its the app name that is also defined in the manifest. 
Screenshot 2023-10-20 at 4 44 17 PM

**Step 2) Add build flavors to iOS**

My learning here is that this requires more steps than android and also relies more on navigating XCode UI (not a quick text edit of a file). In summary, it required creating a new scheme, duplicating build configurations and renaming (ex. "Debug-dev", "Release-dev", "Profile-dev"). Connecting dev build configurations to the dev scheme and prod configurations to the prod scheme. 

This article was the best to follow along because of the gifs showing how to navigate XCode: https://itnext.io/flutter-new-app-setup-with-flavors-in-one-go-331471b127e3

**Step 3) Create 2 entry-points into the app: main_dev.dart and main_prod.dart**
My learning also is that there is no official documentation from Flutter on how to best configure dart code to provide multiple flavors. The flutter docs here only describe setting it up natively (https://docs.flutter.dev/deployment/flavors).

I used flutter_flavors package and followed along with their code descriptions. 
Screenshot 2023-10-20 at 4 38 59 PM

**Step 4) Add configurations to launch.json to allow Visual Studio to build the different flavors.** (This is optional - you can always just build from the terminal too but this is easier). 
Screenshot 2023-10-20 at 4 40 26 PM

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -26,6 +26,20 @@
...
+            "type": "dart",
+            "program": "lib/main_dev.dart",
+            "args": ["--flavor", "dev"]
+        },
+        {
+            "name": "prod",
+            "request": "launch",
+            "type": "dart",
+            "program": "lib/main_prod.dart",
+            "args": ["--flavor", "prod"]
```
Comment (cxw320) [inline .vscode/launch.json:42]: This is for easier use to launch in VS Code.

File: lib/main_prod.dart [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  });
+  runApp(const ProviderScope(child: MyApp()));
+}
+
+class MyApp extends StatelessWidget {
+  const MyApp({super.key});
+
+  // This widget is the root of your application.
+  @override
+  Widget build(BuildContext context) {
```
Comment (bcarterterp) [inline lib/main_prod.dart:20]: I suggest abstracting MyApp out. That way the main* files only need to pass what is important to the widget.

File: pubspec.lock [hunk 3]
```diff
@@ -259,6 +259,14 @@ packages:
     description: flutter
     source: sdk
     version: "0.0.0"
+  flutter_flavor:
```
Comment (bcarterterp) [inline pubspec.lock:262]: I was looking through the dependency's repository, and it didn't seem like it was giving us much. I'd like to know your thoughts.

**Review by bcarterterp — COMMENTED** [review]
Awesome work. One thing I would suggest is treating the Flavors/Environment variables as a dependency. To follow our architecture, you would have something like:
```
EnvironmentVariableRepository
EnvironmentVariableRepositoryImpl
```
Then use them like you would any other dependency. Let me know if you want to take some time and walk through it!

File: pubspec.lock (hunk 3 above)
Comment (cxw320) [reply pubspec.lock:262]: You know what...I just looked through it and you're right. They doing the least :/....

They give us the banner in the corner. And the list of "variables" - which seems to be the most meaningful thing provided for us as part of the FlavorConfig. 

In the [flutter docs](https://docs.flutter.dev/deployment/flavors), they only officially endorse the below 2 packages to use:
Screenshot 2023-10-25 at 10 45 20 AM


flutter_flavorizr does do more - its attempting to automatically generate the main dart files, for example. But I'm concerned by how many issues and PRs people have submitted to this - it makes me think that there are some bugs still being worked out... so I went with flutter_flavor to try to save some minimal boiler plate code.

Screenshot 2023-10-25 at 10 42 28 AM

File: lib/main_dev.dart [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+import 'package:flap_app/presentation/navigation.dart';
+import 'package:flap_app/presentation/theme/app_theme.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter_flavor/flutter_flavor.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+
+void main() {
+  FlavorConfig(name: "dev", location: BannerLocation.topEnd, variables: {
+    "appTitle": "Boiler Plate - Dev Flavor",
+    "baseUrlHost": "api.spoonacular.com" //replace this baseUrl host used for your staging API
```
Comment (amandaroa) [inline lib/main_dev.dart:11]: What if there were multiple host urls depending on the feature? Is this the place to define that or could we do it at the API repository or Implemenation level?

**Comment by amandaroa** [comment]
I like the simplicity of not having a bunch of extra files generated like for image assets and launch screen, however, I really like the pattern and testing you established in https://github.com/bcarterterp/Flap/pull/13. Is there a way to use flutter_flavor for some of the setup but bring over the flavor provider pattern from the other PR?

**Comment by cxw320** [comment]
Closing for now since we determined a route to go without using any package for setting up flavors.

**Pull request closed without merging**